File: old java code/src/GaiaNet/GaiaNet/Node.java
package GaiaNet.GaiaNet;

import GaiaNet.Common.ByteTools;
import GaiaNet.Encrypt.Digest;

import java.io.IOException;
import java.io.Serializable;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Properties;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class Node implements Serializable {
    public int uid;
    public int gid;
    public ArrayList<String> macArray;
    public String innerIp;
    public String outerIp;
    public String ip6;
    public String nodeName;
    public String hostName;
    public String osName;

    // Three methods to build node.
    // 1.local  2.from txt info. 3. null
    public Node(String s, Config conf){
        this(s);
        this.nodeName = conf.nodeName;
    }

    public Node(String s){
        if ("local".equalsIgnoreCase(s)) {
            this.getIPAndHost();
            this.getMacArray();
            this.osName = getOsName();
            this.uid = getUid();
            this.gid = getGid();
        } else {
            buildFromTxt(s);
        }
    }

    public Node(){}

    private void getMacArray(){
        try {
            ArrayList<String> tmpMacList = new ArrayList<>();
            java.util.Enumeration<NetworkInterface> netInterfaceEnum = NetworkInterface.getNetworkInterfaces();
            while (netInterfaceEnum.hasMoreElements()){
                NetworkInterface netIface = netInterfaceEnum.nextElement();
                byte[] mac = netIface.getHardwareAddress();
                if (mac == null) continue;
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < mac.length; i++) {
                    sb.append(String.format("%02X%s", mac[i], (i < mac.length - 1) ? "-" : ""));
                }
                tmpMacList.add(sb.toString());
                this.macArray = tmpMacList;
            }

            if (this.macArray == null) { //Can not read MAC address, I don't know why?
                tmpMacList.add("AB-CD-EF-GH-IJ");
                this.macArray = tmpMacList;
//                System.out.println("The MAC address can not be null");
//                System.exit(0);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    private String getOsName(){
        Properties props=System.getProperties(); //获得系统属性集
        return props.getProperty("os.name");
    }

    private void getIPAndHost(){
        try {
            InetAddress inetAddress = InetAddress.getLocalHost();
            this.innerIp = inetAddress.getHostAddress();
            this.hostName = inetAddress.getHostName();
            this.ip6 = Inet6Address.getLocalHost().getHostAddress();
        } catch (UnknownHostException e) {
            e.printStackTrace();
        }
    }

    private int getUid(){
        int i, res=0;
        try {
            int bitSize = 16;
            String mac = this.macArray.get(0);
            byte[] macHash = Digest.PearsonHash(mac.getBytes(StandardCharsets.UTF_8), bitSize);
            for (i=0; i<macHash.length; i++){
                res += Byte.toUnsignedInt(macHash[i]) * Math.pow(2, 8*(macHash.length-1-i));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return res;
    }


    private int getGid(){
        return 0;
    }

    public String getInfo(){
        String s = "";
        s += "NodeName: " + this.nodeName + "\n";
        s += "Uid: " + this.uid + "\n";
        s += "Gid: " + this.gid + "\n";
        s += "Mac: " + this.macArray.toString() + "\n";
        s += "HostName: " + this.hostName + "\n";
        s += "InnerIP: " + this.innerIp + "\n";
        s += "IP6Addr: " + this.ip6 + "\n";
        s += "OuterIP: " + this.outerIp + "\n";
        s += "osName: " + this.osName + "\n";
        return s;
    }

    @Override
    public String toString(){
        return getInfo();
    }

    public String toTxtForm(){
        String res = String.format("%d  %s  %s%n", this.uid, this.nodeName, this.outerIp);
        return res;
    }

    public void buildFromTxt(String txt){

    }



}

File: GaiaNets/Node.cs
using System;
using System.Net;
using System.Collections.Generic;
using System.Linq;
using System.Net.NetworkInformation;

namespace GaiaNet.GaiaNets
{
    public class Node {
        public int uid;
        public int gid;
        public List<String> macArray = new List<string>();
        public String outerIp;
        public List<IPAddress> ipv4;
        public List<IPAddress> ipv6;
        public String nodeName;
        public String hostName;
        public String osName;
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(
            System.Reflection.MethodBase.GetCurrentMethod().ReflectedType);

        // Three methods to build node.
        // 1.local  2.from txt info. 3. null
        public Node(String s){
            if ("local".Equals(s)) {
                this.GetIPAndHost();
                this.GetMacArray();
                this.osName = Environment.OSVersion.ToString();
                this.uid = GetUid();
                this.gid = GetGid();
                this.nodeName = Config.nodeName;
                log.Info("Node initialize: " + this.ToString());
            } else {
                BuildFromTxt(s);
            }
        }

        public Node(){}

        private void GetMacArray(){
            try {
                this.macArray = NetworkInterface
                    .GetAllNetworkInterfaces()
                    .Where( nic => nic.OperationalStatus == OperationalStatus.Up && nic.NetworkInterfaceType != NetworkInterfaceType.Loopback )
                    .Where( nic => nic.NetworkInterfaceType == NetworkInterfaceType.Ethernet || 
                        nic.NetworkInterfaceType == NetworkInterfaceType.GigabitEthernet  || 
                        nic.NetworkInterfaceType == NetworkInterfaceType.Wireless80211)
                    .Select( nic => nic.GetPhysicalAddress().ToString() )
                    .ToList();

                if (this.macArray == null) { //Can not read MAC address, I don't know why?
                    // tmpMacList.add("AB-CD-EF-GH-IJ");
                    // this.macArray = tmpMacList;
                    log.Error("The MAC address can not be null");
                }

            } catch (Exception e) {
                log.Error(e);
            }
        }

        private void GetIPAndHost(){
            try {
                this.hostName = Dns.GetHostName();
                IPAddress[] ipaddress = Dns.GetHostAddresses(this.hostName);
                this.ipv4 = ipaddress.Where(ip => ip.AddressFamily==System.Net.Sockets.AddressFamily.InterNetwork).ToList();
                this.ipv6 = ipaddress.Where(ip => ip.AddressFamily== System.Net.Sockets.AddressFamily.InterNetworkV6 &&
                                !ip.IsIPv6LinkLocal).ToList();
            } catch (Exception e) {
                log.Error(e);
            }
        }

        private int GetUid(){
            int i, res=0;
            try {
                // int bitSize = 16;
                // String mac = this.macArray.get(0);
                // byte[] macHash = Digest.PearsonHash(mac.getBytes(StandardCharsets.UTF_8), bitSize);
                // for (i=0; i<macHash.length; i++){
                //     res += Byte.toUnsignedInt(macHash[i]) * Math.pow(2, 8*(macHash.length-1-i));
                // }
            } catch (Exception e) {
                log.Error(e);
            }
            return res;
        }


        private int GetGid(){
            return 0;
        }

        public String GetInfo(){
            String s = "\n";
            s += "\t- NodeName: " + this.nodeName + "\n";
            s += "\t- Uid: " + this.uid + "\n";
            s += "\t- Gid: " + this.gid + "\n";
            s += "\t- Mac: " + String.Join(", ", this.macArray) + "\n";
            s += "\t- HostName: " + this.hostName + "\n";
            s += "\t- IP4Addr: " + String.Join(", ", this.ipv4.Select(ip=>ip.ToString())) + "\n";
            s += "\t- IP6Addr: " + String.Join(", ", this.ipv6.Select(ip=>ip.ToString())) + "\n";
            s += "\t- OuterIP: " + this.outerIp + "\n";
            s += "\t- osName: " + this.osName + "\n";
            return s;
        }

        public override String ToString(){
            return GetInfo();
        }

        public String ToTxtForm(){
            String res = String.Format("%d  %s  %s%n", this.uid, this.nodeName, this.outerIp);
            return res;
        }

        public void BuildFromTxt(String txt){

        }

        public byte[] ToBytes(){
            return new byte[10];
        }



    }
}